### ISSUE TITLE
 Don't trigger WordPress.PHP.DevelopmentFunctions if inside 'WP_DEBUG' conditional

### ISSUE THREAD

**Issue by dingo-d** [WordPress-Coding-Standards/WordPress-Coding-Standards#1232]
If you use `error_log()` or `print_r()` functions inside your code, or in my case `error_log( print_r( , true ) );`, the warning will be made in the sniff results

> error_log() found. Debug code should not normally be used in production.

Which is a totally valid warning. But if you are using this

```php
if ( defined( 'WP_DEBUG' ) && WP_DEBUG === true ) {
  error_log( print_r( $some_var, true ) );
}
```

I don't think that the warning should be shown. Since you will only print or send to error log if you are in a debug mode, and that implies that you want this intentionally.

Any thoughts about expanding the `WordPress.PHP.DevelopmentFunctions` sniff?

**Comment by GaryJones** [comment]
I don't think your example is a common pattern, that would make this exception worthwhile.

In my code, I'll often have a different constant that is specific to the theme or plugin. Others may do `true === WP_DEBUG`, or only use `==`, or do it as part of a ternary condition. In terms of the function calls, folks' code may be more OOP (so something like `$this->log->error(...);`) rather than directly calling `error_log`, even if the conditional is consistent.

So, while the overall context makes sense ("not reporting presumably intentional debugging code"), I don't think the practical aspect of covering all useful variations makes adding it a good option.

**Comment by dingo-d** [comment]
Hmm right, haven't thought of the various ways one can add error reporting.

I've looked through the whitelist comment list but didn't notice the appropriate one for this use case.

**Comment by jrfnl** [comment]
@dingo-d As long as you use PHPCS 3.2.0 or higher, you can use the new modular ignore syntax provided by PHPCS for this:

```php
if ( defined( 'WP_DEBUG' ) && WP_DEBUG === true ) {
  // phpcs:disable WordPress.PHP.DevelopmentFunctions
  error_log( print_r( $some_var, true ) );
  // phpcs:enable
}
```

You can even use the precise error code to just and only ignore that particular message instead of the whole sniff. For more information, see: https://github.com/squizlabs/PHP_CodeSniffer/wiki/Advanced-Usage#ignoring-parts-of-a-file

Does that help ?

**Comment by dingo-d** [comment]
Sure thing, this will prove useful. Thanks!
